### PR TITLE
Add `Partner.countries`

### PR DESCRIPTION
**Pull request by willdch**
https://seed-company-squad.monday.com/boards/3451697530/pulses/5037868084

@bryanjnelson here is that draft PR for you to look at.

┆Issue is synchronized with this [Monday item](https://seed-company-squad.monday.com/boards/3451697530/pulses/5124319112) by [Unito](https://www.unito.io)

### CODE REVIEW

**Comment by CarsonF** [comment]
parter?

File: src/components/partner/partner.resolver.ts [hunk 1]
```diff
@@ -93,6 +95,28 @@ export class PartnerResolver {
...
+  async countries(
+    @Parent() partner: Partner,
+    @Loader(LocationLoader) loader: LoaderOf<LocationLoader>,
+  ): Promise<SecuredLocations> {
+    const countries = (await loader.loadMany(partner.countries.value)).flatMap(
+      (country) => {
+        if (country instanceof NotFoundException) {
+          return [];
+        } else if (country instanceof Error) {
+          return country;
```
Comment (bryanjnelson) [inline src/components/partner/partner.resolver.ts:108]: In this case `country` is actually an `Error`, so you want to throw it instead of return it:

```typescript
throw country;
```

This will take care of your build error.

File: src/components/location/dto/location.dto.ts [hunk 2]
```diff
@@ -54,6 +55,11 @@ export class Location extends Resource {
 })
 export class SecuredLocation extends SecuredProperty(Location) {}
 
+@ObjectType({
+  description: SecuredPropertyList.descriptionFor('a location or null'),
```
Comment (bryanjnelson) [inline src/components/location/dto/location.dto.ts:59]:
```suggestion
  description: SecuredPropertyList.descriptionFor('a list of locations'),
```

File: src/components/partner/dto/create-partner.dto.ts [hunk 3]
```diff
@@ -1,8 +1,9 @@
-import { Field, InputType, ObjectType } from '@nestjs/graphql';
+import { Field, ID, InputType, ObjectType } from '@nestjs/graphql';
```
Comment (bryanjnelson) [inline src/components/partner/dto/create-partner.dto.ts:1]: Carson corrected me here...let's alias this instead of the other `ID`.

```suggestion
import { Field, ID as IDType, InputType, ObjectType } from '@nestjs/graphql';
```

File: src/components/partner/dto/create-partner.dto.ts [hunk 4]
```diff
@@ -1,8 +1,9 @@
...
 import { Transform, Type } from 'class-transformer';
 import { Matches, ValidateNested } from 'class-validator';
 import { uniq } from 'lodash';
-import { ID, IdField, IdOf, NameField } from '../../../common';
+import { ID as CordID, IdField, IdOf, NameField } from '../../../common';
```
Comment (bryanjnelson) [inline src/components/partner/dto/create-partner.dto.ts:5]: Let's keep this as the simple `ID` name.

```suggestion
import { ID, IdField, IdOf, NameField } from '../../../common';
```

File: src/components/partner/dto/create-partner.dto.ts [hunk 5]
```diff
@@ -41,6 +42,10 @@ export abstract class CreatePartner {
 
   @IdField({ nullable: true })
   readonly languageOfWiderCommunicationId?: IdOf<Language> | null;
+
+  @Field(() => [ID], { nullable: true })
+  @Transform(({ value }) => uniq(value))
+  readonly countries?: ReadonlyArray<IdOf<Location>> = [];
```
Comment (bryanjnelson) [inline src/components/partner/dto/create-partner.dto.ts:48]:
```suggestion
  @Field(() => [ID], { nullable: true })
  @IsId({ each: true })
  @Transform(({ value }) => uniq(value))
  readonly countries?: ReadonlyArray<IdOf<Location>> = [];
```

File: src/components/partner/dto/update-partner.dto.ts
Comment (bryanjnelson) [inline src/components/partner/dto/update-partner.dto.ts:1]: You can apply the same changes here to `UpdatePartner` as shown above for `CreatePartner`.

File: src/components/partner/partner.module.ts
Comment (bryanjnelson) [inline src/components/partner/partner.module.ts:1]: You can revert these changes...we no longer need them since we'll be 🪓 ing `listCountries` in the service.

File: src/components/partner/partner.repository.ts [hunk 6]
```diff
@@ -103,6 +105,15 @@ export class PartnerRepository extends DtoRepository<
             .raw('WHERE size(projList) = 0')
             .return(`'High' as sensitivity`),
         )
+        .subQuery('node', (sub) =>
+          sub
+            .optionalMatch([
```
Comment (bryanjnelson) [inline src/components/partner/partner.repository.ts:110]:
```suggestion
            .match([
```

File: src/components/partner/partner.repository.ts [hunk 7]
```diff
@@ -103,6 +105,15 @@ export class PartnerRepository extends DtoRepository<
...
+              node('node'),
+              relation('out', '', 'countries', ACTIVE),
```
Comment (bryanjnelson) [inline src/components/partner/partner.repository.ts:112]:
```suggestion
              relation('out', '', 'countries'),
```

File: src/components/partner/partner.service.ts [hunk 8]
```diff
@@ -33,6 +40,7 @@ import { PartnerRepository } from './partner.repository';
 @Injectable()
 export class PartnerService {
   constructor(
+    private readonly countriesService: LocationService,
```
Comment (bryanjnelson) [inline src/components/partner/partner.service.ts:43]: 🪓

File: src/components/partner/partner.service.ts [hunk 9]
```diff
@@ -192,6 +236,23 @@ export class PartnerService {
...
+      session,
+    );
+
+    return {
+      ...countriesListOutput,
+      canRead: true,
+      canCreate: this.privileges.for(session, IProject).can('create'),
+    };
+  }
+
```
Comment (bryanjnelson) [inline src/components/partner/partner.service.ts:255]: 🪓 - no longer needed

File: src/components/partner/partner.service.ts [hunk 10]
```diff
@@ -136,14 +145,49 @@ export class PartnerService {
...
+          return;
+        }
+        throw new NotFoundException(
+          errMsg,
+          `partner.${resourceField}${Array.isArray(ids) ? `[${index}]` : ''}`,
+        );
+      }),
+    );
+  }
+
```
Comment (bryanjnelson) [inline src/components/partner/partner.service.ts:190]: For this whole section you'll be using an abstraction now that handles most of this; you're code will be much simpler.  You can model it after [the example here](https://github.com/SeedCompany/cord-api-v3/blob/develop/src/components/partner/partner.service.ts#L146C2-L158C2).  Note that you'll have to rebase on `develop` to get Carson's abstraction code before you can make these changes.

**Review by bryanjnelson — CHANGES_REQUESTED** [review]
Sorry for all the churn here; this was basically all errors that I made that you inherited.  😬 

Thanks for the flexibility @willdch!

**Review by bryanjnelson — APPROVED** [review]
Thanks @willdch!  Look'n good!

@CarsonF - on to you for a final pass.

**Comment by bryanjnelson** [comment]
> Looks good!
> 
> I wonder if we should add a restriction to only allow Locations that are type == Country?

Sounds reasonable.  At what point would that be?  A custom input validation in the Resolver?  I would think not, because we'd have to make another call to the db to check there.  Perhaps rather in the service layer, on the create and update?

**Comment by CarsonF** [comment]
Yeah the service layer. Could do something like
```ts
const loader = await resourceLoader.getLoader(Location);
const locations = await loadManyIgnoreMissingThrowAny(loader, ids);
const invalid = locations.flatMap(l => l.type !== 'Country' ? l : []);
if (invalid.length) {
	const ex = new InvalidReferencesException('Only locations that are countries can be used', invalid);
	throw ex.withField('partner.countries');
}
```

**Comment by bryanjnelson** [comment]
Cool cool.  Do we really want `InvalidReferencesException` though?  In it's current form it only takes one param.  Unless you want us to alter or overload it to take a second?  I thought perhaps maybe you were meaning a different Exception altogether?

**Comment by CarsonF** [comment]
On second thought we should have a custom exception class for this specific use case. Since classes correspond to codes, that's how a consumer should programmatically check for this error.
You can copy paste the InvalidReferencesException and just change the class name and message.

**Review by bryanjnelson — APPROVED** [review]
LGTM!

@CarsonF ?

**Review by CarsonF — APPROVED** [review]
